data_model/src/error.rs

```diff
@@ -0,0 +1,243 @@
+//! Errors used in Iroha special instructions and
+//! queries. Instruction execution should fail with a specific
+//! error variant, as opposed to a generic (printable-only)
+//! [`eyre::Report`]. If it is impossible to predict what kind of
+//! error shall be raised, there are types that wrap
+//! [`eyre::Report`].
+#![cfg(feature = "std")]
+
+use derive_more::Display;
+use iroha_crypto::HashOf;
+use iroha_primitives::fixed::FixedPointOperationError;
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use thiserror::Error;
+
+use crate::{block::VersionedCommittedBlock, metadata, permission, prelude::*, trigger};
+
+/// Query errors.
+#[derive(Error, Debug, Decode, Encode, IntoSchema)]
+pub enum QueryError {
+    /// Query cannot be decoded.
+    #[error("Query cannot be decoded")]
+    Decode(#[from] Box<iroha_version::error::Error>),
+    /// Query has wrong signature.
+    #[error("Query has the wrong signature: {0}")]
+    Signature(String),
+    /// Query is not allowed.
+    #[error("Query is not allowed: {0}")]
+    Permission(permission::validator::DenialReason),
+    /// Query has wrong expression.
+    #[error("Query has a malformed expression: {0}")]
+    Evaluate(String),
+    /// Query found nothing.
+    #[error("Query found nothing: {0}")]
+    Find(#[from] Box<FindError>),
+    /// Query found wrong type of asset.
+    #[error("Query found wrong type of asset: {0}")]
+    Conversion(String),
+    /// Query without account.
+    #[error("Unauthorized query: account not provided")]
+    Unauthorized,
+}
+
+/// Instruction execution error type
+#[derive(Debug, Error)]
+pub enum IsiError {
```

Comment:
this one we don't use in `client`?

Reply:
Yes, but it uses `FindError` which is also used by `QueryError`,  so i thought it would make sense to group them.

Reply:
```suggestion
pub enum InstructionExecutionFailure {
```

Similarly, `QueryError` -> `QueryExecutionFailure`. 

These names are much more informative and specific. We *know* that the `enum` is an `Error` because it has `#[derive(Error)` on top of it. I don't recommend hiding more things behind feature flags if the purpose is dead code elimination, given that LTO is a thing.

Reply:
@Erigara I believe this type could replace `InstructionExecutionFail::reason` type

Reply:
Created #3298